### DISCUSSION TITLE
ruff config in pyproject.toml

### DISCUSSION THREAD

**Discussion by tvannahl** [mason-org/mason.nvim#1792] [Q&A] [answered]
Moin,

`ruff` is a fantastic tool with great defaults. Unfortunately there is one default, I do like to deviate from and typically I use the `pyproject.toml` for this configuration.

e.g. 

```toml
[tool.ruff]
line-length = 80
```

Now if I execute `ruff format ..` directly in the command line, the `pyproject.toml` is automatically detected and formatting is configured accordingly. But this does not seem to be the case for the `ruff` meson plugin.

Does anyone have an Idea where I would need to start for such a configuration?

**Comment by 0gust1** [comment]
Same concern here. Worse, I prefer tabs than spaces... (put the guns down, please :) )

That said, I'm not completely sure if it's Mason's duties.

**Reply by tvannahl** [reply, marked as answer]
No judgment for as long as its formatted automatically ;-).

I've been able to solve the issue for me just now. Hoping it helps you as well:

1. It's not `mason.vim` who is doing the formatting. This is done by [conform.nvim](https://github.com/stevearc/conform.nvim/tree/master/lua/conform/formatters) instead. You can check the formatter configuration using `:ConformInfo`.
2. There are more formatters than just one `ruff` formatter. I've configured most of them using the following configuration:


```lua
return {
  {
    "stevearc/conform.nvim",
    opts = {
      formatters_by_ft = {
        python = { "ruff", "ruff_organize_imports", "ruff_format" },
      },
    },
  },
}
```

**Reply by 0gust1** [reply]
great, thanks! I can confirm that the formatting now respects the project local config.

...Now, my concern was with linting :D
=> I kept having warnings "W191 indentation contains tabs".
=> it turns out that I used `ruff_lsp` that is now deprecated in favor of `ruff`

**Reply by tvannahl** [reply]
That's great. You're right with `ruff_lsp` vs `ruff` - it seemed to have been integrated recently.

On the `W191` note, I guess adding the following to your `pyproject.toml` should fix that warning for you:

```
[tool.ruff.lint]
ignore = ["W191"]
```

[source](https://docs.astral.sh/ruff/settings/#lintpydocstyle)

**Comment by simlmx** [comment]
I had to use this for the `ruff` to use my `pyproject.toml` config:

```lua
return {
  {
    "stevearc/conform.nvim",
    opts = {
      formatters_by_ft = {
        python = { "ruff_fix", "ruff_format" },
      },
    },
  },
}
```
